rdr_service/offline/sync_consent_files.py

```diff
@@ -42,6 +45,117 @@ class ParticipantPairingInfo:
+
+    def check_consents(self, files: Collection[ConsentFile], session):
+        # Get the latest pairing information for the participants
+        participant_ids = [file.participant_id for file in files]
+        raw_pairing_data = self._participant_history_dao.get_pairing_history(
+            session=self._session,
+            participant_ids=participant_ids
+        )
+        latest_participant_pairings = {}
+        for history_record in raw_pairing_data:
+            participant_id = history_record.participantId
+            possible_latest_record = PairingHistoryRecord(
+                org_name=history_record.externalId,
+                start_date=history_record.lastModified
+            )
+            if participant_id not in latest_participant_pairings:
+                latest_participant_pairings[participant_id] = possible_latest_record
+            else:
+                currently_stored_pairing = latest_participant_pairings[participant_id]
+                if currently_stored_pairing.start_date <= possible_latest_record.start_date \
+                        and currently_stored_pairing.org_name != possible_latest_record.org_name:
+                    latest_participant_pairings[participant_id] = possible_latest_record
+
+        participant_summaries = ParticipantSummaryDao.get_by_ids_with_session(self._session, participant_ids)
+        for summary in participant_summaries:
+            pairing_info = latest_participant_pairings[summary.participantId]
+            latest_participant_pairings[summary.participantId] = (pairing_info, summary)
+
+        count = 0
+        for file in files:
+            pairing_info, summary = latest_participant_pairings[file.participant_id]
+            sync_date = self.get_sync_date(file=file, summary=summary, latest_pairing_info=pairing_info)
+            if sync_date is not None:
+                file.sync_time = sync_date
+                file.sync_status = ConsentSyncStatus.SYNC_COMPLETE
+
+            count += 1
+            if count % 500 == 0:
+                session.commit()
+
+    @classmethod
+    def get_sync_date(cls, file: ConsentFile, summary: ParticipantSummary, latest_pairing_info: PairingHistoryRecord):
+        """
+        This encapsulates the code for determining if a consent would have been copied using the previous process
+        for syncing the files. It checks to see if a file would have been available for sync based on any of the times
+        that a sync would have been triggered by the data available for a participant.
+        If any of them would have triggered a sync to the latest org's bucket, then the date is returned for the sync.
+        Otherwise None is returned.
+        """
+        if not cls._org_date_valid_for_sync(
+            upload_time=file.file_upload_time,
+            latest_pair_time=latest_pairing_info.start_date
+        ):
+            return None
+
+        primary_time = summary.consentForStudyEnrollmentTime
```

Comment:
So this tool only has to worry about the last received primary / EHR consents?    (For cases where there was also a previous "first yes" / multiple consent files of that type submitted)

Reply:
We only receive the Primary consent once. For EHR: I've tried to take the simple route and have it detect whether files would have transferred based on EHR time for the latest submission. This may miss detecting that some participants have had files copied over for times that another EHR consent submission would have caused it, but in those scenarios we would simply copy their files over again. I think that's an acceptable outcome.

Reply:
Can you clarify?  We certainly have participants with multiple ConsentPII payloads (although unclear if they would reference the same pdf?), and we have the consent updates for earlier cohorts for the primary consent.   I wasn't sure if those cases were part of what's being checked here?